File: src/improvement/doc_drift.py
"""Documentation-code drift detection.

Detects mismatches between documentation and code:
- Function signatures that changed but docs weren't updated
- Module renames not reflected in docs
- Features added/removed without doc updates
- Stale file path references
"""

from __future__ import annotations

import ast
import os
import re
from dataclasses import dataclass


@dataclass
class DriftFinding:
    doc_file: str
    line: int
    category: str  # "signature", "path", "module", "feature"
    description: str


def find_code_references(doc_content: str) -> list[tuple[int, str]]:
    """Extract code references from documentation.

    Returns list of (line_number, reference) tuples.
    References include: file paths, function names in backticks, module names.
    """
    refs = []
    for i, line in enumerate(doc_content.split("\n"), 1):
        # File paths (e.g., `src/engine/state.py`)
        for match in re.finditer(r'`([a-zA-Z0-9_/.-]+\.py)`', line):
            refs.append((i, match.group(1)))
        # Function references (e.g., `validate_plan()`)
        for match in re.finditer(r'`([a-zA-Z_][a-zA-Z0-9_]*)\(\)`', line):
            refs.append((i, match.group(1) + "()"))
    return refs


def check_path_references(
    doc_file: str,
    doc_content: str,
    project_dir: str,
) -> list[DriftFinding]:
    """Check if file path references in a doc still exist."""
    findings = []
    refs = find_code_references(doc_content)

    for line_num, ref in refs:
        if ref.endswith("()"):
            continue  # Function ref, not path ref
        full_path = os.path.join(project_dir, ref)
        if not os.path.exists(full_path):
            findings.append(DriftFinding(
                doc_file=doc_file,
                line=line_num,
                category="path",
                description=f"Referenced path `{ref}` does not exist",
            ))

    return findings


def check_function_references(
    doc_file: str,
    doc_content: str,
    project_dir: str,
) -> list[DriftFinding]:
    """Check if function references in a doc still exist in the codebase."""
    findings = []
    refs = find_code_references(doc_content)

    # Build a set of all function names in the project
    all_functions: set[str] = set()
    for root, dirs, files in os.walk(project_dir):
        dirs[:] = [d for d in dirs if d not in {"__pycache__", ".git", "node_modules", ".venv"}]
        for name in files:
            if not name.endswith(".py"):
                continue
            filepath = os.path.join(root, name)
            try:
                with open(filepath) as f:
                    tree = ast.parse(f.read())
                for node in ast.walk(tree):
                    if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
                        all_functions.add(node.name)
            except (SyntaxError, OSError):
                continue

    for line_num, ref in refs:
        if not ref.endswith("()"):
            continue
        func_name = ref[:-2]  # Remove ()
        if func_name not in all_functions:
            findings.append(DriftFinding(
                doc_file=doc_file,
                line=line_num,
                category="signature",
                description=f"Referenced function `{func_name}()` not found in codebase",
            ))

    return findings


def scan_doc(
    doc_file: str,
    project_dir: str,
) -> list[DriftFinding]:
    """Scan a single doc file for drift."""
    try:
        with open(doc_file) as f:
            content = f.read()
    except (FileNotFoundError, OSError):
        return []

    findings = []
    findings.extend(check_path_references(doc_file, content, project_dir))
    findings.extend(check_function_references(doc_file, content, project_dir))
    return findings


def scan_docs_dir(
    docs_dir: str,
    project_dir: str,
) -> list[DriftFinding]:
    """Scan all documentation files for drift."""
    findings = []

    if not os.path.isdir(docs_dir):
        return findings

    for name in sorted(os.listdir(docs_dir)):
        if not name.endswith(".md"):
            continue
        doc_path = os.path.join(docs_dir, name)
        findings.extend(scan_doc(doc_path, project_dir))

    return findings

File: rust/src/improvement/doc_drift.rs
//! Documentation-code drift detection — stale file paths and function references.

use regex::Regex;
use serde::{Deserialize, Serialize};
use std::path::Path;

/// A documentation drift finding.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DriftFinding {
    pub doc_file: String,
    pub line: usize,
    pub category: String,
    pub description: String,
}

/// Extract code references from documentation content.
///
/// Returns `(line_number, reference)` tuples for file paths and function names.
pub fn find_code_references(doc_content: &str) -> Vec<(usize, String)> {
    let path_re = Regex::new(r"`([a-zA-Z0-9_/.\-]+\.py)`").unwrap();
    let func_re = Regex::new(r"`([a-zA-Z_][a-zA-Z0-9_]*)\(\)`").unwrap();

    let mut refs = Vec::new();
    for (i, line) in doc_content.lines().enumerate() {
        let line_num = i + 1;
        for cap in path_re.captures_iter(line) {
            refs.push((line_num, cap[1].to_string()));
        }
        for cap in func_re.captures_iter(line) {
            refs.push((line_num, format!("{}()", &cap[1])));
        }
    }
    refs
}

/// Check if file path references in a doc still exist.
pub fn check_path_references(
    doc_file: &str,
    doc_content: &str,
    project_dir: &str,
) -> Vec<DriftFinding> {
    let refs = find_code_references(doc_content);
    let mut findings = Vec::new();

    for (line_num, reference) in refs {
        if reference.ends_with("()") {
            continue;
        }
        let full_path = Path::new(project_dir).join(&reference);
        if !full_path.exists() {
            findings.push(DriftFinding {
                doc_file: doc_file.to_string(),
                line: line_num,
                category: "path".into(),
                description: format!("Referenced path `{}` does not exist", reference),
            });
        }
    }
    findings
}

/// Check if function references in a doc exist in a known function set.
pub fn check_function_references(
    doc_file: &str,
    doc_content: &str,
    known_functions: &std::collections::HashSet<String>,
) -> Vec<DriftFinding> {
    let refs = find_code_references(doc_content);
    let mut findings = Vec::new();

    for (line_num, reference) in refs {
        if !reference.ends_with("()") {
            continue;
        }
        let func_name = &reference[..reference.len() - 2];
        if !known_functions.contains(func_name) {
            findings.push(DriftFinding {
                doc_file: doc_file.to_string(),
                line: line_num,
                category: "signature".into(),
                description: format!(
                    "Referenced function `{}()` not found in codebase",
                    func_name
                ),
            });
        }
    }
    findings
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashSet;

    #[test]
    fn test_find_code_references_paths() {
        let content = "See `src/engine/state.py` for details.";
        let refs = find_code_references(content);
        assert_eq!(refs.len(), 1);
        assert_eq!(refs[0].1, "src/engine/state.py");
    }

    #[test]
    fn test_find_code_references_functions() {
        let content = "Call `validate_plan()` to check.";
        let refs = find_code_references(content);
        assert_eq!(refs.len(), 1);
        assert_eq!(refs[0].1, "validate_plan()");
    }

    #[test]
    fn test_check_function_references_missing() {
        let content = "Use `nonexistent_func()` here.";
        let known = HashSet::new();
        let findings = check_function_references("doc.md", content, &known);
        assert_eq!(findings.len(), 1);
        assert_eq!(findings[0].category, "signature");
    }

    #[test]
    fn test_check_function_references_found() {
        let content = "Use `my_func()` here.";
        let mut known = HashSet::new();
        known.insert("my_func".to_string());
        let findings = check_function_references("doc.md", content, &known);
        assert!(findings.is_empty());
    }
}
